### PR TITLE
docs(perf-test) add docs for the performance testing framework

### PR DESCRIPTION
**Pull request by fffonion**
### Review

@reviewer

### CODE REVIEW

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+It allows Kong developers and users to evaluate performance of Kong itself,
+bundled or custom plugins, as well as plot framegraphs to debug performance bottleneck.
+The framework collects RPS (request per second) and latencies of Kong processing the request
+to represent performance metrics under different workloads.
+
+The framework is implemented as an extension to Kong's intergration test suite.
+
+## Installation
+
+The frameworks uses `busted` and some Lua development dependencies of Kong. To setup the environment,
```
Comment (PaulFischer-Kong) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:17]: Add a link to the busted testing framework when referencing 'busted' here. So a user to learn more about the framework.

Comment (fffonion) [reply app/gateway-oss/2.5.x/performance-testing-framework.md]: Added!

**Review by kikito — APPROVED** [review]

**Comment by PaulFischer-Kong** [comment]
Would it be possible to include a section which describes the 'output' one could expect from running these performance tests?

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+Set retry time for each “driver" operation. By default every operation is retried 3 times.
+
+[Back to top](#introduction)
+
+### perf.setup
+
+*syntax: helpers = perf.setup()*
+
+Prepare environment and returns the spec.helpers module. Throws error if any.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:183]:
```suggestion
Prepares environment and returns the `spec.helpers` module. Throws error if any.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+The framework setup some environment variables before importing spec.helpers modules.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:185]:
```suggestion
The framework sets up some environment variables before importing `spec.helpers` modules.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+The returned helpers is just a nomal spec.helpers module, user can use same pattern
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:186]:
```suggestion
The returned helper is just a normal `spec.helpers` module, user can use same pattern
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+in integration tests to setup entities in Kong. DB-less mode is currently not implemented.
+
+[Back to top](#introduction)
+
+### perf.start_upstream
+
+*syntax: upstream_uri = perf.start_upstream(nginx_conf_blob)*
+
+Start upstream (Nginx/OpenResty) with given nginx_conf_blob. Returns the upstream
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:195]:
```suggestion
Starts upstream (Nginx/OpenResty) with given `nginx_conf`_blob. Returns the upstream
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+URI accessible from Kong's view. Throws error if any.
+
+[Back to top](#introduction)
+
+### perf.start_kong
+
+*syntax: perf.start_kong(version, kong_configs?)*
+
+Start Kong with given version and Kong configurations in kong_configs as a Lua table.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:204]:
```suggestion
Starts Kong with given version and Kong configurations in `kong_configs` as a Lua table.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+Stops Kong. Throws error if any.
+
+[Back to top](#introduction)
+
+### perf.teardown
+
+*syntax: perf.teardown(full?)*
+
+Teardown. Throws error if any. With terraform driver, set full to true terminates
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:225]:
```suggestion
Teardown. Throws error if any. With the terraform driver, setting full to true terminates
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+all infrastructure while by default it does cleanup only to speed up successive runs.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:226]:
```suggestion
all infrastructure, while by default it does cleanup only to speed up successive runs.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+[Back to top](#introduction)
+
+### perf.start_stapxx
+
+*syntax: perf.start_stapxx(stapxx_file_name, arg?)*
+
+Start the Stap++ script with stapxx_file_name exists in   and addtional CLI args args. Throws error if any.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:234]:
```suggestion
Starts the Stap++ script with `stapxx_file_name` exists in and additional CLI args. Throws error if any.
```

Comment (HCloward) [reply app/gateway-oss/2.5.x/performance-testing-framework.md:234]: Not sure what this means, "Start the Stap++ script with `stapxx_file_name` exists in and additional CLI args" - is something missing here?

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+
+This function blocks until the SystemTap module is fully prepared and inserted into
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:236]:
```suggestion
This function blocks until the `SystemTap` module is fully prepared and inserted into the
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+kernel. It should be called before perf.start_load.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:237]:
```suggestion
kernel. It should be called before `perf.start_load`.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+
+[Back to top](#introduction)
+
+### perf.start_load
+
+*syntax: perf.start_load(options?)*
+
+Start to send load to Kong using wrk. Throws error if any. options is a Lua table that may contains:
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:245]:
```suggestion
Starts to send load to Kong using `wrk`. Throws error if any. Options is a Lua table that may contain the following:
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+
+- **path** string request path, default to /
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:247]:
```suggestion
- **path** String request path; defaults to `/ `.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+[Back to top](#introduction)
+
+### perf.start_load
+
+*syntax: perf.start_load(options?)*
+
+Start to send load to Kong using wrk. Throws error if any. options is a Lua table that may contains:
+
+- **path** string request path, default to / 
+- **uri** base URI except path, default to http://kong-ip:kong-port/
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:248]:
```suggestion
- **uri** Base URI exception path; defaults to `http://kong-ip:kong-port/`.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+- **connections** connection count, default to 1000
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:249]:
```suggestion
- **connections** Connection count; defaults to 1000.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+- **threads** request thread count, default to 5
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:250]:
```suggestion
- **threads** Request thread count; defaults to 5. 
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+
+*syntax: perf.start_load(options?)*
+
+Start to send load to Kong using wrk. Throws error if any. options is a Lua table that may contains:
+
+- **path** string request path, default to / 
+- **uri** base URI except path, default to http://kong-ip:kong-port/
+- **connections** connection count, default to 1000
+- **threads** request thread count, default to 5 
+- **duration** number perf test duration in seconds, default to 10
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:251]:
```suggestion
- **duration** Number of performance tests duration in seconds; defaults to 10. 
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+*syntax: perf.start_load(options?)*
+
+Start to send load to Kong using wrk. Throws error if any. options is a Lua table that may contains:
+
+- **path** string request path, default to / 
+- **uri** base URI except path, default to http://kong-ip:kong-port/
+- **connections** connection count, default to 1000
+- **threads** request thread count, default to 5 
+- **duration** number perf test duration in seconds, default to 10 
+- **script** content of wrk script as string, default to nil
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:252]:
```suggestion
- **script** Content of wrk script as string; defaults to nil.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+
+[Back to top](#introduction)
+
+### perf.wait_result
+
+*syntax: result = perf.start_load(options?)*
+
+Wait the load test to finish and return the result as string. Throws error if any.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:260]:
```suggestion
Waits for the load test to finish and returns the result as a string. Throws error if any.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+
+Currently this function waits indefintely until both wrkand stap++ process to exit.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:262]:
```suggestion
Currently this function waits indefinitely, until both wrk and stap++ process to exit.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+
+[Back to top](#introduction)
+
+### perf.combine_results
+
+*syntax: combined_result = perf.combine_results(results, …)*
+
+Calculate multiple results returned by perf.wait_result and return their average and min/max.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:270]:
```suggestion
Calculates multiple results returned by `perf.wait_result` and returns their average and min/max.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+Throws error if any.
+
+[Back to top](#introduction)
+
+### perf.generate_flamegraph
+
+*syntax: perf.generate_flamegraph(path, title?)*
+
+Generate a flamegraph with title and save to path. Throws error if any.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:279]:
```suggestion
Generates a FlameGraph with title and saves to path. Throws error if any.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+
+[Back to top](#introduction)
+
+### perf.save_error_log
+
+*syntax: perf.save_error_log(path)*
+
+Save Kong error log to path. Throws error if any.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:287]:
```suggestion
Saves Kong error log to path. Throws error if any.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+
+[Back to top](#introduction)
+
+## Customization
+
+### Add new test suite
+
+All tests are stored in `spec/04-perf/01-rps` and `spec/04-perf/02-flamegraph` of Kong repository.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:295]:
```suggestion
All tests are stored in `spec/04-perf/01-rps` and `spec/04-perf/02-flamegraph` of the Kong repository.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
+Add a new file under one of the directory and put `#tags` in test description.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:296]:
```suggestion
Add a new file under one of the directories and put `#tags` in the test description.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 26]
```diff
@@ -0,0 +1,321 @@
...
+
+[Back to top](#introduction)
+
+### Add new provider in terraform
+
+Users can use the terraform driver in most major service providers as long as
+it's supported by terraform. The following contracts are made between the framework and terraform module:
+
+The terraform files are stored in `spec/fixtures/perf/terraform/<provider>`
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:305]:
```suggestion
The terraform files are stored in `spec/fixtures/perf/terraform/<provider>`.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 27]
```diff
@@ -0,0 +1,321 @@
...
+
+Two instances are provisioned, one for running Kong and another for running upstream
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:307]:
```suggestion
Two instances are provisioned, one for running Kong and another for running an upstream
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 28]
```diff
@@ -0,0 +1,321 @@
...
+and load (worker). Firewall allows bidirectional traffic between the two instances
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:308]:
```suggestion
and load (worker). A firewall allows bidirectional traffic between the two instances
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 29]
```diff
@@ -0,0 +1,321 @@
...
+and from public internet. Both instances run Ubuntu 20.04/focal.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:309]:
```suggestion
and from the public internet. Both instances run Ubuntu 20.04/focal.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 30]
```diff
@@ -0,0 +1,321 @@
...
+
+An SSH key to login into both instances exists or will be created
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:311]:
```suggestion
An SSH key to login into both instances exists or will be created in
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 31]
```diff
@@ -0,0 +1,321 @@
...
+`spec/fixtures/perf/terraform/<provider>/id_rsa`. The logged-in user has root privilege.
+
+Following terraform output variables:
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:314]:
```suggestion
The following are terraform output variables:
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 32]
```diff
@@ -0,0 +1,321 @@
...
+
+- **kong-ip** Kong node public IP
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:316]:
```suggestion
- **kong-ip** Kong node public IP.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 33]
```diff
@@ -0,0 +1,321 @@
...
+- **kong-internal-ip** Kong node internal IP (if unavailable, provide kong-ip)
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:317]:
```suggestion
- **kong-internal-ip** Kong node internal IP (if unavailable, provide kong-ip).
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 34]
```diff
@@ -0,0 +1,321 @@
...
+- **worker-ip** Worker node public IP
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:318]:
```suggestion
- **worker-ip** Worker node public IP.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 35]
```diff
@@ -0,0 +1,321 @@
...
+- **worker-internal-ip** Worker node internal IP (if unavailable, provide worker-ip)
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:319]:
```suggestion
- **worker-internal-ip** Worker node internal IP (if unavailable, provide worker-ip).
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 36]
```diff
@@ -0,0 +1,321 @@
...
+- **tfvars** Terraform variables as a Lua table; for `equinix-metal` provider,
+`packet_project_id` and `packet_auth_token` is required.
+
+[Back to top](#introduction)
+
+### perf.set_log_level
+
+*syntax: perf.set_log_level(level)*
+
+Set the log level, expect one of `"debug"`, `"info"`, `"notice"`, `"warn"`, `"error"` and
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:166]:
```suggestion
Sets the log level; expect one of `"debug"`, `"info"`, `"notice"`, `"warn"`, `"error"` and
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 37]
```diff
@@ -0,0 +1,321 @@
...
+`"crit"`. The default log level is `"info"`.
+
+[Back to top](#introduction)
+
+### perf.set_retry_count
+
+*syntax: perf.set_retry_count(count)*
+
+Set retry time for each “driver" operation. By default every operation is retried 3 times.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:175]:
```suggestion
Sets retry time for each “driver" operation. By default every operation is retried 3 times.
```

**Review by HCloward — COMMENTED** [review]
Made some suggestions about wording.

**Comment by fffonion** [comment]
@HCloward Thank you for the detailed review of this PR!
@PaulFischer-Kong Yes I will add an sample output.

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 38]
```diff
@@ -0,0 +1,362 @@
...
+It allows Kong developers and users to evaluate the performance of Kong itself as well as 
+bundled or custom plugins, and plot framegraphs to debug performance bottlenecks.
+The framework collects RPS (request per second) and latencies of Kong processing the request
+to represent performance metrics under different workloads.
+
+The framework is implemented as an extension to Kong's integration test suite.
+
+## Installation
+
+The frameworks uses [busted](https://olivinelabs.com/busted/) and some
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:17]:
```suggestion
The framework uses [busted](https://olivinelabs.com/busted/) and some
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 39]
```diff
@@ -0,0 +1,362 @@
...
+Lua development dependencies of Kong. To setup the environment,
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:18]:
```suggestion
Lua development dependencies of Kong. To set up the environment,
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 40]
```diff
@@ -0,0 +1,362 @@
...
+| Driver   | Test between git commits | Test between binary releases | Flamegraph | Test unreleased version |
+|----------|--------------------------|------------------------------|------------|-------------------------|
+| local    | yes                      |                              | yes        | yes                     |
+| docker   | yes                      | yes                          |            |                         |
+|terraform | yes                      | yes                          | yes        |                         |
+
+- **local** Driver reuses users' local environment. It's faster to run,
+but the RPS and latency number may be influenced by other local programs and thus inaccurate.
+
+    * Requires Lua development dependencies of Kong, OpenResty and wrk to be installed.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:37]:
```suggestion
    * Requires Lua development dependencies of Kong, OpenResty, and `wrk` be installed.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 41]
```diff
@@ -0,0 +1,362 @@
...
+    * Requires SystemTap, kernel headers and build chain to be installed if generating flamegraph.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:38]:
```suggestion
    * Requires SystemTap, kernel headers, and build chain to be installed if generating FlameGraph.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 42]
```diff
@@ -0,0 +1,362 @@
...
+as it requires less dependencies. But it may also be influenced by other local programs
+and Docker network performance. And it doesn't support FlameGraph generation.
+
+- **terraform** Driver runs in remote VM or bare metal machine. It's most accurate,
+but it requires Terraform knowledge to operate and setup.
+
+    * Requires the [Terraform](https://www.terraform.io/downloads.html) binary be installed.
+    * Requires git binary if testing between git commits. When testing between git commits,
+    the framework assumes the current directory is Kong's repo. It will stash your working
+    directory and unstash after test is finished. When using docker or terraform driver,
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:50]:
```suggestion
    directory and unstash after test is finished. When using the docker or terraform driver,
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 43]
```diff
@@ -0,0 +1,362 @@
...
+    the framework derives the base version of each commit and uses matching Docker image or
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:51]:
```suggestion
    the framework derives the base version of each commit and uses the matching Docker image or
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 44]
```diff
@@ -0,0 +1,362 @@
...
+    Kong binary package and pours local source code inside.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:52]:
```suggestion
    Kong binary package and puts local source code inside.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 45]
```diff
@@ -0,0 +1,362 @@
...
+[Back to top](#introduction)
+
+## API
+
+### perf.use_driver
+
+*syntax: perf.use_driver(name, options?)*
+
+Uses driver name, which must be one of "local", "docker"  or  "terraform". Additional
+parameters for driver can be specified in options as a Lua table. Throws error if any.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:193]:
```suggestion
parameters for the driver can be specified in options as a Lua table. Throws error if any.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 46]
```diff
@@ -0,0 +1,362 @@
...
+
+Only terraform driver expects options parameter, which contains following keys:
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:195]:
```suggestion
Only the terraform driver expects an options parameter, which contains following keys:
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 47]
```diff
@@ -0,0 +1,362 @@
...
+[Back to top](#introduction)
+
+### perf.setup
+
+*syntax: helpers = perf.setup()*
+
+Prepares environment and returns the `spec.helpers` module. Throws error if any.
+
+The framework sets up some environment variables before importing `spec.helpers` modules.
+The returned helper is just a normal `spec.helpers` module, user can use same pattern
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:227]:
```suggestion
The returned helper is just a normal `spec.helpers` module. Users can use the same pattern
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 48]
```diff
@@ -0,0 +1,362 @@
...
+Teardown. Throws error if any. With the terraform driver, setting full to true terminates
+all infrastructure, while by default it does cleanup only to speed up successive runs.
+
+[Back to top](#introduction)
+
+### perf.start_stapxx
+
+*syntax: perf.start_stapxx(stapxx_file_name, arg?)*
+
+Starts the Stap++ script with `stapxx_file_name` exists in and additional CLI args. Throws error if any.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:275]: Not sure about "exists in" part of this sentence.

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 49]
```diff
@@ -0,0 +1,362 @@
...
+
+This function blocks until the `SystemTap` module is fully prepared and inserted into the
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:277]: Blocks what?

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 50]
```diff
@@ -0,0 +1,362 @@
...
+*syntax: perf.start_load(options?)*
+
+Starts to send load to Kong using `wrk`. Throws error if any. Options is a Lua table that may contain the following:
+
+- **path** String request path; defaults to `/ `.
+- **uri** Base URI exception path; defaults to `http://kong-ip:kong-port/`.
+- **connections** Connection count; defaults to 1000.
+- **threads** Request thread count; defaults to 5. 
+- **duration** Number of performance tests duration in seconds; defaults to 10. 
+- **script** Content of wrk script as string; defaults to nil.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:293]:
```suggestion
- **script** Content of `wrk` script as string; defaults to nil.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 51]
```diff
@@ -0,0 +1,362 @@
...
+
+[Back to top](#introduction)
+
+### perf.wait_result
+
+*syntax: result = perf.start_load(options?)*
+
+Waits for the load test to finish and returns the result as a string. Throws error if any.
+
+Currently this function waits indefinitely, until both wrk and stap++ process to exit.
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:303]:
```suggestion
Currently, this function waits indefinitely, or until both `wrk` and Stap++ processes exit.
```

File: app/gateway-oss/2.5.x/performance-testing-framework.md [hunk 52]
```diff
@@ -0,0 +1,362 @@
...
+
+All tests are stored in `spec/04-perf/01-rps` and `spec/04-perf/02-flamegraph` of the Kong repository.
+Add a new file under one of the directories and put `#tags` in the test description.
+
+[Back to top](#introduction)
+
+### Add new provider in terraform
+
+Users can use the terraform driver in most major service providers as long as
+it's supported by terraform. The following contracts are made between the framework and terraform module:
```
Comment (HCloward) [inline app/gateway-oss/2.5.x/performance-testing-framework.md:344]:
```suggestion
it's supported by Terraform. The following contracts are made between the framework and terraform module:
```

**Review by HCloward — APPROVED** [review]
Couple more suggestions - otherwise LGTM - thanks @fffonion !

**Comment by fffonion** [comment]
@HCloward I addresses the two parts in question, let me know it that clarifies the description : )

**Review by HCloward — APPROVED** [review]
Thanks @fffonion - looks great!

**Review by PaulFischer-Kong — APPROVED** [review]

**Pull request merged**